rust/xaynet-mobile/src/ffi/participant.rs

```diff
@@ -0,0 +1,355 @@
+///       to train a model, and set it once the
+///       [`PARTICIPANT_SHOULD_SET_MODEL`] flag is set.
+///     - [`PARTICIPANT_SHOULD_SET_MODEL`]: if set, then the
+///       participant should set its model, by calling
+///       [`xaynet_ffi_participant_set_model()`]
+///
+/// # Safety
+///
+/// When calling this method, you have to ensure that *either* the
+/// pointer is NULL *or* all of the following is true:
+///
+/// - The pointer must be properly [aligned].
+/// - It must be "dereferencable" in the sense defined in the
+///   [`::std::ptr`] module documentation.
+///
+/// After destroying the participant with
+/// [`xaynet_ffi_participant_destroy`] becomes invalid and must not be
+/// used.
+///
+/// [`::std::ptr`]: https://doc.rust-lang.org/std/ptr/index.html#safety
+/// [aligned]: https://doc.rust-lang.org/std/ptr/index.html#alignment
+#[no_mangle]
+pub unsafe extern "C" fn xaynet_ffi_participant_tick(participant: *mut Participant) -> c_int {
+    let participant = match unsafe { participant.as_mut() } {
+        Some(participant) => participant,
+        None => return ERR_NULLPTR,
+    };
+
+    participant.tick();
+
+    let mut flags: c_int = 0;
+    match participant.task() {
+        Task::None => flags |= PARTICIPANT_TASK_NONE,
+        Task::Sum => flags |= PARTICIPANT_TASK_SUM,
+        Task::Update => flags |= PARTICIPANT_TASK_UPDATE,
+    };
+    if participant.should_set_model() {
+        flags |= PARTICIPANT_SHOULD_SET_MODEL;
+    }
+    if participant.made_progress() {
+        flags |= PARTICIPANT_MADE_PROGRESS;
+    }
+    flags
+}
+
+/// Serialize the participant state and return a buffer that contains
+/// the serialized participant.
+///
+/// # Safety
+///
+/// 1. When calling this method, you have to ensure that *either* the
+///    pointer is NULL *or* all of the following is true:
+///      - The pointer must be properly [aligned].
+///      - It must be "dereferencable" in the sense defined in the
+///        [`::std::ptr`] module documentation.
+/// 2. the `ByteBuffer` created by this function must be destroyed
+///    with [`xaynet_ffi_participant_destroy`]. Attempting to free the
+///    memory from the other side of the FFI is UB.
+///
+/// [`::std::ptr`]: https://doc.rust-lang.org/std/ptr/index.html#safety
+/// [aligned]: https://doc.rust-lang.org/std/ptr/index.html#alignment
+///
+/// # Example
+///
+/// To save the participant into a file:
+///
+/// ```c
+///  const ByteBuffer *save_buf = xaynet_ffi_participant_save(participant);
+///  assert(save_buf);
+///
+///  char *path = "./participant.bin";
+///  FILE *f = fopen(path, "w");
+///  fwrite(save_buf->data, 1, save_buf->len, f);
+///  fclose(f);
+/// ```
+#[no_mangle]
+pub unsafe extern "C" fn xaynet_ffi_participant_save(
+    participant: *mut Participant,
+) -> *const ByteBuffer {
+    let participant: Participant = match unsafe { participant.as_mut() } {
+        Some(ptr) => unsafe { *Box::from_raw(ptr) },
```

Comment:
Yeah the problem is that `save()` currently consumes the participant... I think it could be changed, but for now it's a limitation.

I should definitely mention that in the "Safety" section though: the pointer becomes invalid.